Nested dune call fix (execution environment part)

Leonidas-from-XIV opened the pull request:
This is a companion PR to #12902. Where #12902 resolves the `dune` binary to be called to be correct, this PR prepends an extra entry to the `PATH` variable so the environment will point to the right dune instead of whatever was captured in the environment (potentially by accident).

It has a non-essential dependency on #12916 to show that the Dune binary is correctly resolved in all cases.

Leonidas-from-XIV committed:
Prepend dune to path when running commands in Pkg

Signed-off-by: Marek Kubica <marek@tarides.com>

Leonidas-from-XIV committed:
Update test to show that it doesn't capture the fake dune anymore

Signed-off-by: Marek Kubica <marek@tarides.com>

Leonidas-from-XIV commented on src/dune_rules/run_with_path.ml:
```diff
@@ -141,11 +141,21 @@ module Spec = struct
           |> String.concat ~sep:"")
       in
       let metadata = Process.create_metadata ~purpose:ectx.metadata.purpose () in
+      let dune_folder =
```
I wanted to avoid creating a random folder in `_build` to avoid managing its state. But I can create an action that creates it, that way it is managed via the build system. Does this sound like a better solution to you?

rgrinberg replied:
That’s what happens with opam when dune changes. It is the right thing to do because a new version of dune might interpret the rules differently so they would need to be re-executed.

if you’re worried about the recompilation being slow, I think it’s just a matter of turning on the shared cache for the sub process dunes that we launch.

rgrinberg replied:
How about instead of looking at the action you just look at the package dependencies of the package that corresponds to this action. If the package dependencies contain dune, then we should introduce this dependency. This way we don't need look at the action itself at all.

This way, you can also skip doing this setup for packages that do not depend on dune.

Leonidas-from-XIV committed:
Add dependency on dune path to action

Signed-off-by: Marek Kubica <marek@tarides.com>

rgrinberg commented on src/dune_rules/run_with_path.ml:
```diff
@@ -120,11 +124,11 @@ module Spec = struct
              | Path p -> path p)))
     in
-    List [ List ([ prog ] @ args); path ocamlfind_destdir ]
+    List [ List ([ prog ] @ args); path ocamlfind_destdir; path dune_exe ]
```
It is the content of the executable that's important here. Whether it's in /usr/bin or /usr/local/bin (to give an example) should make no difference.

I think you've done the right thing with `Action_builder.path` already though. This is all that is required and the modification to the action's own digest is what gives us this undesired sensitivity to the path.

Leonidas-from-XIV committed:
Track whether the package had a dune dependency

We don't register the dependency because we will not build it but we add
it to the type to know whether the package depended on Dune.

Signed-off-by: Marek Kubica <marek@tarides.com>

Leonidas-from-XIV committed:
Add dependency on Dune in test packages

Signed-off-by: Marek Kubica <marek@tarides.com>

Leonidas-from-XIV commented on test/blackbox-tests/test-cases/pkg/different-dune-in-path.t:
```diff
@@ -2,22 +2,42 @@ Clarify the behavior when the `dune` in PATH is not the one used to start the bu
   $ make_test_package foo
+  $ mkpkg foo <<EOF
+  > depends: [ "dune" ]
+  > build: [ "dune" "build" "-p" name "@install" ]
```
Undid the change. Turns out I only needed to add a `(depends dune)` stanza.

rgrinberg committed:
refactor: some cosmetic changes

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>
